server/src/main/java/org/tctalent/server/service/db/impl/DuolingoExamServiceImpl.java

```diff
@@ -0,0 +1,131 @@
+        if (isIdenticalExamExists(existingExams, newScore, newYear, newNotes)) {
+          LogBuilder.builder(log)
+              .action("UpdateCandidateExams")
+              .message(String.format("Identical Duolingo exam record already exists for candidate ID: %s. Skipping...", candidateOpt.getId()))
+              .logInfo();
+        }
+        else {
+          // Create and save new exam record
+          createAndSaveNewExam(candidateOpt, newScore, newYear, newNotes);
+        }
+      }
+    }
+  }
+
+  private boolean isIdenticalExamExists(List<CandidateExam> existingExams, String newScore, String newYear, String newNotes) {
+    return existingExams.stream().anyMatch(existingExam ->
+        Exam.DuolingoOfficial.equals(existingExam.getExam()) &&
+            Objects.equals(newScore, existingExam.getScore().trim()) &&
+            Objects.equals(newYear, String.valueOf(existingExam.getYear()).trim()) &&
+            Objects.equals(newNotes, existingExam.getNotes())
+    );
+  }
+
+  private void createAndSaveNewExam(Candidate candidate, String newScore, String newYear, String newNotes) {
+    CreateCandidateExamRequest candidateExamRequest = new CreateCandidateExamRequest();
+    candidateExamRequest.setExam(Exam.DuolingoOfficial);
+    candidateExamRequest.setScore(newScore);
+    candidateExamRequest.setYear(Long.valueOf(newYear));
+    candidateExamRequest.setNotes(newNotes);
+
+    candidateExamService.createExam(candidate.getId(), candidateExamRequest);
+    LogBuilder.builder(log)
+        .action("UpdateCandidateExams")
+        .message(String.format("New Duolingo exam record added for candidate ID: %s.", candidate.getId()))
+        .logInfo();
+  }
+
+  private String buildNotes(DuolingoDashboardResponse result) {
+    return String.format(
+        "Generated from Duolingo dashboard | Certificate URL: %s | Interview URL: %s | Verification Date: %s | Percent Score: %s | "
+            + "Scale: %s | Literacy Subscore: %s | Conversation Subscore: %s | Comprehension Subscore: %s | Production Subscore: %s",
+        result.getCertificate_url(), result.getInterview_url(), result.getVerification_date(),
+        result.getPercent_score(), result.getScale(), result.getLiteracy_subscore(),
+        result.getConversation_subscore(), result.getComprehension_subscore(),
+        result.getProduction_subscore());
+  }
+
+  /**
+   * Extracts the year from a date string in the format "yyyy-MM-dd".
+   *
+   * @param dateStr the input date string
+   * @return the year as a Long, or null if the input is invalid
+   */
+  private Long extractYear(String dateStr) {
+    try {
+      LocalDate date = LocalDate.parse(dateStr);
+      return (long) date.getYear();
+    } catch (DateTimeParseException e) {
+      // Log and handle error
+      LogBuilder.builder(log)
+          .action("extractYear")
+          .message(String.format("Invalid date format: %s", dateStr))
+          .logError();
+      return null;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Swallowing exceptions without logging can make it difficult to identify and troubleshoot issues, potentially leading to application instability, hidden bugs, or unexpected behavior. To avoid this, ensure that all exceptions are logged using appropriate logging levels. When logging exceptions, include the exception object or its stack trace to capture detailed information. For critical exceptions, rethrow them when necessary to provide sufficient context for effective troubleshooting.